frontend/src/components/grantaccess/GrantAccess.vue

```diff
@@ -0,0 +1,393 @@
+import {
+    selectedApplication,
+    selectedApplicationDisplayText,
+} from '@/store/ApplicationState';
+import {
+    grantAccessFormData,
+    domainOptions,
+    FOREST_CLIENT_INPUT_MAX_LENGTH,
+    setGrantAccessFormData,
+    getGrantAccessFormData,
+} from '@/store/GrantAccessDataState';
+
+import type { FamApplicationRole, FamForestClient } from 'fam-api';
+
+const defaultFormData = {
+    domain: domainOptions.IDIR,
+    userId: '',
+    forestClientNumber: '',
+    role_id: null as number | null,
+};
+const formData = ref(JSON.parse(JSON.stringify(defaultFormData))); // clone default input
+const formValidationSchema = object({
+    userId: string()
+        .required('User ID is required')
+        .min(2, 'User ID must be at least 2 characters')
+        .nullable(),
+    role_id: number().required('Please select a value'),
+    forestClientNumber: string()
+        .when('role_id', {
+            is: (_role_id: number) => isAbstractRoleSelected(),
+            then: () =>
+                string()
+                    .required('Forest Client number is required')
+                    .min(8, 'Forest Client ID must be 8 characters long'),
+        })
+        .nullable(),
+});
+
+const loading = ref<boolean>(false);
+const applicationRoleOptions = ref<FamApplicationRole[]>([]);
+const forestClientData = ref<FamForestClient[] | null>(null);
+const invalidForestClient = ref<boolean>(false);
+
+const apiServiceFactory = new ApiServiceFactory();
+const applicationsApi = apiServiceFactory.getApplicationApi();
+const forestClientApi = apiServiceFactory.getForestClientApi();
+
+onMounted(async () => {
+    try {
+        applicationRoleOptions.value = (
+            await applicationsApi.getFamApplicationRoles(
+                selectedApplication.value?.application_id as number
+            )
+        ).data as FamApplicationRole[];
+
+        if (grantAccessFormData.value) {
+            formData.value = getGrantAccessFormData();
+
+            invalidForestClient.value = isAbstractRoleSelected();
+        } else {
+            resetForm();
+        }
+    } catch (error: unknown) {
+        router.push('/dashboard');
+        return Promise.reject(error);
+    }
+});
+
+function userIdChange() {
+    resetForestClientNumberData();
+}
+
+const getSelectedRole = (): FamApplicationRole | undefined => {
+    return applicationRoleOptions.value.find(
+        (item) => item.role_id === formData.value.role_id
+    );
+};
+
+const isAbstractRoleSelected = () => {
+    return getSelectedRole()?.role_type_code == 'A';
+};
+
+function forestClientCheckOnlyDigit(evt: KeyboardEvent) {
+    if (isNaN(parseInt(evt.key))) {
+        evt.preventDefault();
+    }
+}
+
+function resetForestClientNumberData() {
+    invalidForestClient.value = true;
+    forestClientData.value = null;
+}
+
+function resetForm() {
+    formData.value = defaultFormData;
+}
+
+function cancelForm() {
+    resetForm();
```

Comment:
this still doesn't look correct to me..... when we click on the cancel, do we want to reset the form to be empty? then in this case, we might need to reset the grantAccessFormData in the store as well, otherwise whenever I click grant access again, it will populate the stored grantAccessFormData, then the reset here doesn't make any sense.